### PR TITLE
Full hybrid support

### LINKED ISSUE #3: Add hybrid support
Both for KEMs (#16) as well as for SIGs (quite some upstream integration points missing; possibly requiring some X.509/encode wizardry)

### LINKED ISSUE #6: Implement interoperable key gen & storage
Goal: Use provider to run `openssl genpkey -algorithm <SIG> -out <SIG>_srv.key` such that it outputs the same file as the `OQS-OpenSSL1_1_1` fork. 
This depends on #2 and #3.

### LINKED ISSUE #7: Implement interoperable X509 read/write
Goal: Execute provider-based `openssl [x509] req -new -newkey <SIG> [...]` such that it creates [X509] files interoperable with the `OQS-OpenSSL1_1_1` fork.

This depends on #2 and #3.

### PR DESCRIPTION
**Pull request by baentsch**
Fixes #3 :
- Updates data structures and logic for hybrid keys to also support signatures
- Implements hybrid signatures
- Adds tests for hybrid signatures

Fixes #6 :
Fixes #7 :
- Adds interoperability tests with oqs-openssl111 for X509 structures (keys, certs) for all supported key types

@bhess: Please note the changes as to how classic key/sig lengths get stored in the `privkey` and `pubkey` structures and how the `comp_*` keys are being used: There are some slight semantics changes aimed at re-using and retaining (most of) the hybrid KEM logic you added while also marshalling data for encoding (and consequent decoding).

@dstebila For now this code maintains & tests full interoperability with oqs-openssl. Once this code gets merged, my [concern raised here](https://github.com/open-quantum-safe/openssl/issues/333#issuecomment-931909690) is dropped and we could begin implementing https://github.com/open-quantum-safe/openssl/issues/333 (and/or #39)

### CODE REVIEW

**Review by bhess — CHANGES_REQUESTED** [review]
Looks good, I added a few comments and suggestions.

File: oqsprov/oqs_kmgmt.c [hunk 1]
```diff
@@ -692,7 +909,7 @@ static void *sphincsshake256128frobust_gen_init(void *provctx, int selection)
                                                       \
     static void *ecp_##tokalg##_gen_init(void *provctx, int selection) \
     { \
-        return oqsx_gen_init(provctx, selection, tokoqsalg, NULL /* TBD: create hybrid TLS alg name! */, KEY_TYPE_ECP_HYB_KEM, bit_security); \
+        return oqsx_gen_init(provctx, selection, tokoqsalg, NULL, KEY_TYPE_ECP_HYB_KEM, bit_security); \
```
[the remark this answers is not in the document]
Comment (baentsch) [reply oqsprov/oqs_kmgmt.c:912]: That (tls_)name is only necessary for name->OID mapping. And KEMs don't have OIDs AFAIK.

File: oqsprov/oqs_kmgmt.c [hunk 2]
```diff
@@ -724,7 +941,7 @@ static void *sphincsshake256128frobust_gen_init(void *provctx, int selection)
                                                       \
     static void *ecx_##tokalg##_gen_init(void *provctx, int selection) \
     { \
-        return oqsx_gen_init(provctx, selection, tokoqsalg, NULL /* TBD: create hybrid TLS alg name! */, KEY_TYPE_ECX_HYB_KEM, bit_security); \
+        return oqsx_gen_init(provctx, selection, tokoqsalg, NULL, KEY_TYPE_ECX_HYB_KEM, bit_security); \
```
[the remark this answers is not in the document]
Comment (baentsch) [reply oqsprov/oqs_kmgmt.c:944]: That (tls_)name is only necessary for name->OID mapping. And KEMs don't have OIDs AFAIK.

File: oqsprov/oqs_sig.c [hunk 3]
```diff
@@ -252,62 +192,223 @@ static int oqs_sig_verify_init(void *vpoqs_sigctx, void *voqssig, const OSSL_PAR
...
-        ERR_raise(ERR_LIB_USER, OQSPROV_R_INVALID_SIZE);
-        return 0;
+    if (is_hybrid) {
+        if ((classical_ctx_sign = EVP_PKEY_CTX_new(evpkey, NULL)) == NULL ||
+            EVP_PKEY_sign_init(classical_ctx_sign) <= 0) {
+          ERR_raise(ERR_LIB_USER, ERR_R_FATAL);
+          goto endsign;
+        }
+        if (oqsxkey->evp_info->keytype == EVP_PKEY_RSA) {
+            if (EVP_PKEY_CTX_set_rsa_padding(classical_ctx_sign, RSA_PKCS1_PADDING) <= 0) {
```
[the remark this answers is not in the document]
Comment (baentsch) [reply oqsprov/oqs_sig.c:242]: Good point -- but again, something to discuss in https://github.com/open-quantum-safe/oqs-provider/issues/32: Reason for this code: [oqs-openssl code](https://github.com/open-quantum-safe/openssl/blob/461e98f8f57b7042558709537a5faf2872f6b61d/crypto/ec/oqs_meth.c#L1485)

File: oqsprov/oqs_decode_der2key.c [hunk 4]
```diff
@@ -586,34 +586,58 @@ static void oqsx_key_adjust(void *key, struct der2key_ctx_st *ctx)
 
 ///// OQS_TEMPLATE_FRAGMENT_DECODER_MAKE_START
 MAKE_DECODER("dilithium2", dilithium2, oqsx, PrivateKeyInfo);
-MAKE_DECODER("dilithium2", dilithium2, oqsx, SubjectPublicKeyInfo);
+MAKE_DECODER("dilithium2", dilithium2, oqsx, SubjectPublicKeyInfo);MAKE_DECODER("p256_dilithium2", p256_dilithium2, oqsx, PrivateKeyInfo);
```
[the remark this answers is not in the document]
Comment (baentsch) [reply oqsprov/oqs_decode_der2key.c]: Good point. Fixed in new commit.

File: oqs-template/oqsprov/oqs_decode_der2key.c/decoder_make.fragment [hunk 5]
```diff
@@ -2,6 +2,10 @@
    {%- for variant in sig['variants'] %}
 MAKE_DECODER("{{ variant['name'] }}", {{ variant['name'] }}, oqsx, PrivateKeyInfo);
 MAKE_DECODER("{{ variant['name'] }}", {{ variant['name'] }}, oqsx, SubjectPublicKeyInfo);
+     {%- for classical_alg in variant['mix_with'] -%}
```
[the remark this answers is not in the document]
Comment (baentsch) [reply oqs-template/oqsprov/oqs_decode_der2key.c/decoder_make.fragment]: Good point. Fixed in new commit.

File: oqsprov/oqs_encode_key2any.c [hunk 6]
```diff
@@ -1006,90 +1074,210 @@ MAKE_ENCODER(dilithium2, oqsx, EncryptedPrivateKeyInfo, pem);
 MAKE_ENCODER(dilithium2, oqsx, PrivateKeyInfo, der);
 MAKE_ENCODER(dilithium2, oqsx, PrivateKeyInfo, pem);
 MAKE_ENCODER(dilithium2, oqsx, SubjectPublicKeyInfo, der);
-MAKE_ENCODER(dilithium2, oqsx, SubjectPublicKeyInfo, pem);
+MAKE_ENCODER(dilithium2, oqsx, SubjectPublicKeyInfo, pem);MAKE_ENCODER(p256_dilithium2, oqsx, EncryptedPrivateKeyInfo, der);
```
[the remark this answers is not in the document]
Comment (baentsch) [reply oqsprov/oqs_encode_key2any.c]: Good point. Fixed in new commit.

File: oqsprov/oqs_encode_key2any.c [hunk 7]
```diff
@@ -548,13 +544,12 @@ static int oqsx_pki_priv_to_der(const void *vecxkey, unsigned char **pder)
     // Encoding private _and_ public key concatenated ... seems unlogical and unnecessary, 
     // but is what oqs-openssl does, so we repeat it for interop... also from a security 
     // perspective not really smart to copy key material (side channel attacks, anyone?),
-    // but so be it.
+    // but so be it for now (TBC).
```
[the remark this answers is not in the document]
Comment (baentsch) [reply oqsprov/oqs_encode_key2any.c:547]: Hence the "TBC" addition. Proposals (also for [the corresponding oqs-openssl code](https://github.com/open-quantum-safe/openssl/blob/461e98f8f57b7042558709537a5faf2872f6b61d/crypto/ec/oqs_meth.c#L1073)) welcome.

File: tmp/p256_dilithium2_CA.crt [hunk 8]
```diff
@@ -0,0 +1,88 @@
+-----BEGIN CERTIFICATE-----
```
Comment (bhess) [inline tmp/p256_dilithium2_CA.crt:1]: Are these tmp files added on purpose?

Comment (baentsch) [reply tmp/p256_dilithium2_CA.crt:1]: Absolutely no. Sorry. Will revert. Artifact of running the interop tests in parallel to the `commit` :-(

**Review by bhess — APPROVED** [review]

**Pull request merged**